### PR TITLE
circuit: optimize hash computation

### PR DESCRIPTION
**Pull request by KeefeL**
### Description

Optimize the hash computation of the transaction.

### Rationale

The variable has 256bits, so we can merge several int64 fields into one variable when computing transaction hash. By this way, we can reduce the circuit constraints.

### Example

NA

### Changes

Notable changes:
* The hash computation of the transaction has been changed.

### CODE REVIEW

File: circuit/types/atomic_match.go [hunk 1]
```diff
@@ -83,38 +83,21 @@ func SetAtomicMatchTxWitness(tx *AtomicMatchTx) (witness AtomicMatchTxConstraint
...
-		tx.AccountIndex,
-		tx.BuyOffer.Type,
-		tx.BuyOffer.OfferId,
-		tx.BuyOffer.AccountIndex,
-		tx.BuyOffer.NftIndex,
-		tx.BuyOffer.AssetId,
-		tx.BuyOffer.AssetAmount,
-		tx.BuyOffer.ListedAt,
-		tx.BuyOffer.ExpiredAt,
+		PackInt64Variables(api, tx.AccountIndex, nonce, expiredAt, ChainId),
```
Comment (guagualvcha) [inline circuit/types/atomic_match.go:89]: Please confirm whether it is ok to fill in all 256 bits(exceed 254)

**Comment by yycen** [comment]
For curve BN254, the variable bitlen is 254, so pack 4 int64 into one could possibly result in overflow.

**Comment by yycen** [comment]
And treat each element equally as 64bit int could be too coarse in some cases, such as tx.type or asset.id. Is it possible to have a more detailed or customized bin packing scheme to potentialy pack more items and also avoid the overflow problem?

File: circuit/types/atomic_match.go (hunk 1 above)
Comment (KeefeL) [reply circuit/types/atomic_match.go]: fixed

**Review by guagualvcha — APPROVED** [review]

**Pull request merged**